python/foglamp/common/statistics.py

```diff
@@ -74,25 +84,37 @@ def __init__(self, storage):
                     raise KeyError
             # If key was not present, add the key and with value = value_increment
             except KeyError:
-                desc_txt = "The number of readings received by FogLAMP since startup for sensor {}".format(key)
-                payload = PayloadBuilder().INSERT(key=key, description=desc_txt,
-                                                  value=value_increment, previous_value=0).payload()
-                self._storage.insert_into_tbl("statistics", payload)
+                _logger.exception('Statistics key %s has not been registered', key)
+                raise
             except Exception as ex:
                 _logger.exception(
                     'Unable to update statistics value based on statistics_key %s and value_increment %s, error %s'
                     , key, value_increment, str(ex))
                 raise
 
-# TODO: FOGL-484 Move below commented code to tests directory
-# async def _main():
-#     _storage = StorageClient(core_management_host="0.0.0.0", core_management_port=33881)
-#
-#     _stats = Statistics(_storage)
-#     await _stats.update(key='READINGS', value_increment=10)
-#
-#
-# if __name__ == '__main__':
-#     import asyncio
-#     loop = asyncio.get_event_loop()
-#     loop.run_until_complete(_main())
+    async def register(self, key, description):
+        if key in self._registered_keys:
+            return
+        if len(self._registered_keys) == 0:
+            self._load_keys()
+        try:
+            payload = PayloadBuilder().INSERT(key=key, description=description, value=0, previous_value=0).payload()
+            self._storage.insert_into_tbl("statistics", payload)
+            self._registered_keys.append(key)
+        except Exception as ex:
+            """ The error may be because the key has been created in another process, reload keys """
+            self._load_keys()
```

Comment:
Will it not be an expensive overhead call? Can't we do it via upsert?
PostgreSQL added the ON CONFLICT (key) target action clause (do update) to the INSERT statement to support the upsert feature.

Reply:
No in both cases. Because the class is a singleton it will happen once per process. So it is not an overhead on ever statistic insertion.

We can do it at the storage layer because it does not have enough information to register the statistic, it does not have the description. The previous workaround used a fixed description that would not be applicable for statistics being added by anything other than a south service. This mechanism is more flexible.